### ISSUE TITLE
r14 globals

### ISSUE THREAD

**Issue by kayceesrk** [ocamllabs/ocaml-multicore#182]
Currently, each domain's local state is stored at the start of the domain's minor heap area. This area is accessed by masking the least significant bits of the allocation pointer (r15 on amd64). A better design is to appropriate the r14 register, currently used for storing the exception pointer to point to the domain local state area and move the exception pointer to the domain local state. This design simplifies some of the code in accessing the domain local state in `amd64.S` and `<arch>/emit.mlp`. Similar change is necessary for other architectures. 

With this change, the performance of pushing and popping exception trap frames will change. The performance change is not expected to be significant. Requires benchmarking. 

This is a candidate for upstreaming. The idea is to introduce the domain local state area for all of the globals littered around the runtime. As a result, all of the globals used in the runtime will reside in one place. 

This change should also make allocations faster?

```
subq  $0x18, %r15
leaq  _caml_young_limit(%rip), %rax
cmpq  (%rax), %r15
jb  0x100001607
```

will become

```
subq $0x18, %r15
cmpq OFFSET_OF_YOUNG_LIMIT(%r14), %r15
jb  0x100001607
```

where `OFFSET_OF_YOUNG_LIMIT` is a statically known offset.

**Comment by stedolan** [comment]
Closed by #215

**Comment by kayceesrk** [comment]
Upstream PR ocaml/ocaml#8713
